test-common/utils.c

```diff
@@ -579,14 +583,201 @@ maybe_hashtable_lookup (GHashTable *table,
+  gboolean has_parent =
+    (commit > 0 &&  g_hash_table_lookup (commit_to_ref, GUINT_TO_POINTER (commit)));
+  guint parent_commit = commit == 0 ? commit : commit - 1;
+
+  if (has_parent)
+    populate_commit_chain (commit_graph, parent_commit, collection_ref, commit_to_ref);
+
+  /* Populate commit info now */
+  eos_test_updater_insert_commit_steal_info (commit_graph,
+                                             parent_commit,
+                                             eos_test_updater_commit_info_new (commit,
+                                                                               parent_commit,
+                                                                               collection_ref));
+}
+
+static gint
+sort_commits (gconstpointer lhs, gconstpointer rhs)
+{
+  /* Technically this is playing games with casting -
+   * we should not have commits larger than G_MAXINT here */
+  gint lhs_commit = GPOINTER_TO_INT (lhs);
+  gint rhs_commit = GPOINTER_TO_INT (rhs);
+
+  /* Descending */
+  return rhs_commit - lhs_commit;
+}
+
+void
+eos_test_updater_populate_commit_graph_from_leaf_nodes (EosTestUpdaterCommitGraph *commit_graph,
+                                                        GHashTable                *leaf_nodes)
+{
+  g_autoptr(GHashTable) commit_to_ref = reverse_hashtable (leaf_nodes,
+                                                           g_direct_hash,
+                                                           g_direct_equal,
+                                                           NULL,
+                                                           (GDestroyNotify) ostree_collection_ref_free,
+                                                           no_copy,
+                                                           (CopyFunc) ostree_collection_ref_dup);
+
+  /* Each of the key-value pairs in leaf_nodes points to a candidate
+   * leaf node for a given refspec. From there we recursively go down the
+   * tree and create new EosTestUpdaterCommitInfo objects, unless we
+   * see an entry for that commit in commit_to_ref (we'll start from
+   * that key the next time around */
+  GList *commit_keys = g_list_sort (g_hash_table_get_keys (commit_to_ref),
+                                    sort_commits);
+  GList *iter = commit_keys;
+
+  /* Clear the hash-table first */
+
+  for (; iter != NULL; iter = iter->next)
+    {
+      guint commit = GPOINTER_TO_UINT (iter->data);
+      const OstreeCollectionRef *collection_ref =
+        g_hash_table_lookup (commit_to_ref, GUINT_TO_POINTER (commit));
+      populate_commit_chain (commit_graph,
+                             commit,
+                             collection_ref,
+                             commit_to_ref);
+    }
+
+  g_list_free (commit_keys);
+}
+
+EosTestUpdaterCommitGraph *
+eos_test_updater_commit_graph_new_from_leaf_nodes (GHashTable *leaf_nodes)
+{
+  EosTestUpdaterCommitGraph *commit_graph = eos_test_updater_commit_graph_new ();
```

Comment:
Why have `…_commit_graph_new()` and `…_commit_graph_new_from_leaf_nodes()` as separate functions? Just have one `…_commit_graph_new (GHashTable *leaf_nodes)` function which does the same as `…_new_from_leaf_nodes()` when called with a non-`NULL` `leaf_nodes`, and does the same as `…_new()` when called with a `NULL` parameter.